In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
sns.set()

In [2]:
raw_data = pd.read_csv("FIFA19.csv")

In [3]:
raw_data.head()

ID          Name  Age                                        Photo  \
0   16   Luis García   37   https://cdn.sofifa.org/players/4/19/16.png   
1   41       Iniesta   34   https://cdn.sofifa.org/players/4/19/41.png   
2   80  E. Belözoğlu   37   https://cdn.sofifa.org/players/4/19/80.png   
3  164      G. Pinzi   37  https://cdn.sofifa.org/players/4/19/164.png   
4  657    D. Vaughan   35  https://cdn.sofifa.org/players/4/19/657.png   

  Nationality                                 Flag  Overall  Potential  \
0       Spain  https://cdn.sofifa.org/flags/45.png       71         71   
1       Spain  https://cdn.sofifa.org/flags/45.png       86         86   
2      Turkey  https://cdn.sofifa.org/flags/48.png       79         79   
3       Italy  https://cdn.sofifa.org/flags/27.png       70         70   
4       Wales  https://cdn.sofifa.org/flags/50.png       66         66   

                    Club                                        Club Logo  \
0              KAS Eupen    https://cdn.sofifa.org/teams/2/light/2013.png   
1            Vissel Kobe  https://cdn.sofifa.org/teams/2/light/101146.png   
2  Medipol Başakşehir FK  https://cdn.sofifa.org/teams/2/light/101014.png   
3                 Padova  https://cdn.sofifa.org/teams/2/light/110912.png   
4           Notts County    https://cdn.sofifa.org/teams/2/light/1937.png   

   ... Composure Marking  StandingTackle SlidingTackle  GKDiving  GKHandling  \
0  ...      79.0    70.0            43.0          40.0       9.0        12.0   
1  ...      89.0    67.0            57.0          56.0       6.0        13.0   
2  ...      80.0    65.0            67.0          60.0       7.0        11.0   
3  ...      68.0    74.0            71.0          66.0       5.0        15.0   
4  ...      82.0    59.0            66.0          62.0       6.0         9.0   

   GKKicking GKPositioning GKReflexes Release Clause  
0       13.0          11.0       11.0          €1.1M  
1        6.0          13.0        7.0         €26.9M  
2        7.0          14.0        8.0          €7.6M  
3        7.0           6.0       13.0          €372K  
4       10.0           9.0        7.0          €263K  

[5 rows x 88 columns]

# Clean up missing and unwanted values

## Clean unwanted columns

In [4]:
raw_data.columns

Index(['ID', 'Name', 'Age', 'Photo', 'Nationality', 'Flag', 'Overall',
       'Potential', 'Club', 'Club Logo', 'Value', 'Wage', 'Special',
       'Preferred Foot', 'International Reputation', 'Weak Foot',
       'Skill Moves', 'Work Rate', 'Body Type', 'Real Face', 'Position',
       'Jersey Number', 'Joined', 'Loaned From', 'Contract Valid Until',
       'Height', 'Weight', 'LS', 'ST', 'RS', 'LW', 'LF', 'CF', 'RF', 'RW',
       'LAM', 'CAM', 'RAM', 'LM', 'LCM', 'CM', 'RCM', 'RM', 'LWB', 'LDM',
       'CDM', 'RDM', 'RWB', 'LB', 'LCB', 'CB', 'RCB', 'RB', 'Crossing',
       'Finishing', 'HeadingAccuracy', 'ShortPassing', 'Volleys', 'Dribbling',
       'Curve', 'FKAccuracy', 'LongPassing', 'BallControl', 'Acceleration',
       'SprintSpeed', 'Agility', 'Reactions', 'Balance', 'ShotPower',
       'Jumping', 'Stamina', 'Strength', 'LongShots', 'Aggression',
       'Interceptions', 'Positioning', 'Vision', 'Penalties', 'Composure',
       'Marking', 'StandingTackle', 'SlidingTackle', 'GKDiv

In [5]:
data=raw_data.copy()
data=data.drop(["Photo", "Flag", "Club", "Club Logo", "Jersey Number", "Real Face", "Joined", "Loaned From", "Contract Valid Until"], axis=1)
data

ID             Name  Age          Nationality  Overall  Potential  \
0          16      Luis García   37                Spain       71         71   
1          41          Iniesta   34                Spain       86         86   
2          80     E. Belözoğlu   37               Turkey       79         79   
3         164         G. Pinzi   37                Italy       70         70   
4         657       D. Vaughan   35                Wales       66         66   
5         768           Felipe   33               Brazil       75         75   
6        1179        G. Buffon   40                Italy       88         88   
7        2147  M. Stekelenburg   35          Netherlands       73         73   
8        2335     A. Wilbraham   38              England       63         63   
9        2702       K. Ellison   39              England       57         57   
10       3281        Tarantini   34             Portugal       75         75   
11       3484        J. O'Shea   37  Republic of Ireland       69         69   
12       3665         B. Nivet   41               France       71         71   
13       5479         Casillas   37                Spain       82         82   
14       6826         G. Barry   37              England       72         72   
15       7647      D. Dainelli   39                Italy       73         73   
16       8205            Braga   35             Portugal       70         70   
17       9014        A. Robben   34          Netherlands       84         84   
18       9833          D. Boer   37          Netherlands       69         69   
19      10466      S. Ishizaki   36               Sweden       67         67   
20      10884         M. Flynn   37                Wales       59         59   
21      10899        C. Porter   34              England       62         62   
22      11381         M. Gilks   36             Scotland       67         67   
23      11430       J. McCombe   35              England       53         53   
24      11800         C. Weale   36              England       57         57   
25      11811         P. Green   35  Republic of Ireland       63         63   
26      12940      E. Pogatetz   35              Austria       68         68   
27      13883      D. Whitaker   37              England       61         61   
28      14907      A. Bizzarri   40            Argentina       76         76   
29      16097       N. Rimando   39        United States       71         71   
...       ...              ...  ...                  ...      ...        ...   
18177  246452        S. Bouzar   18               France       57         70   
18178  246454         D. Revan   17              England       53         72   
18179  246455      Pablo Ramón   17                Spain       60         81   
18180  246456    D. Bajselmani   19          Netherlands       59         70   
18181  246457        C. Ngonge   18              Belgium       63         84   
18182  246462        D. García   21            Argentina       59         68   
18183  246464      J. Theissen   20              Germany       57         68   
18184  246465         Guerrero   16                Spain       63         76   
18185  246466      G. Pintidis   18              Germany       52         68   
18186  246469          A. Urzi   18            Argentina       59         76   
18187  246548      O. Sahraoui   17               Norway       52         67   
18188  246549          M. Siby   21               France       54         61   
18189  246550          G. Haag   18               France       55         72   
18190  246552    O. Sverrisson   25               Sweden       54         57   
18191  246565       B. Diakité   17               France       60         80   
18192  246592          Algobia   19                Spain       60         77   
18193  246593           Javito   17                Spain       56         72   
18194  246594      H. Massengo   16               France       62         75   
18195  2465

In [6]:
data.columns

Index(['ID', 'Name', 'Age', 'Nationality', 'Overall', 'Potential', 'Value',
       'Wage', 'Special', 'Preferred Foot', 'International Reputation',
       'Weak Foot', 'Skill Moves', 'Work Rate', 'Body Type', 'Position',
       'Height', 'Weight', 'LS', 'ST', 'RS', 'LW', 'LF', 'CF', 'RF', 'RW',
       'LAM', 'CAM', 'RAM', 'LM', 'LCM', 'CM', 'RCM', 'RM', 'LWB', 'LDM',
       'CDM', 'RDM', 'RWB', 'LB', 'LCB', 'CB', 'RCB', 'RB', 'Crossing',
       'Finishing', 'HeadingAccuracy', 'ShortPassing', 'Volleys', 'Dribbling',
       'Curve', 'FKAccuracy', 'LongPassing', 'BallControl', 'Acceleration',
       'SprintSpeed', 'Agility', 'Reactions', 'Balance', 'ShotPower',
       'Jumping', 'Stamina', 'Strength', 'LongShots', 'Aggression',
       'Interceptions', 'Positioning', 'Vision', 'Penalties', 'Composure',
       'Marking', 'StandingTackle', 'SlidingTackle', 'GKDiving', 'GKHandling',
       'GKKicking', 'GKPositioning', 'GKReflexes', 'Release Clause'],
      dtype='object')

## Clean unwanted rows based on individual columns

In [7]:
pd.set_option("display.max_columns", None)
data.describe(include="all")

ID          Name           Age Nationality       Overall  \
count    18207.000000         18207  18207.000000       18207  18207.000000   
unique            NaN         17194           NaN         164           NaN   
top               NaN  J. Rodríguez           NaN     England           NaN   
freq              NaN            11           NaN        1662           NaN   
mean    214298.338606           NaN     25.122206         NaN     66.238699   
std      29965.244204           NaN      4.669943         NaN      6.908930   
min         16.000000           NaN     16.000000         NaN     46.000000   
25%     200315.500000           NaN     21.000000         NaN     62.000000   
50%     221759.000000           NaN     25.000000         NaN     66.000000   
75%     236529.500000           NaN     28.000000         NaN     71.000000   
max     246620.000000           NaN     45.000000         NaN     94.000000   

           Potential  Value   Wage       Special Preferred Foot  \
count   18207.000000  18207  18207  18207.000000          18159   
unique           NaN    217    144           NaN              2   
top              NaN  €1.1M    €1K           NaN          Right   
freq             NaN    431   4900           NaN          13948   
mean       71.307299    NaN    NaN   1597.809908            NaN   
std         6.136496    NaN    NaN    272.586016            NaN   
min        48.000000    NaN    NaN    731.000000            NaN   
25%        67.000000    NaN    NaN   1457.000000            NaN   
50%        71.000000    NaN    NaN   1635.000000            NaN   
75%        75.000000    NaN    NaN   1787.000000            NaN   
max        95.000000    NaN    NaN   2346.000000            NaN   

        International Reputation     Weak Foot   Skill Moves       Work Rate  \
count               18159.000000  18159.000000  18159.000000           18159   
unique                       NaN           NaN           NaN               9   
top                          NaN           NaN           NaN  Medium/ Medium   
freq                         NaN           NaN           NaN            9810   
mean                    1.113222      2.947299      2.361308             NaN   
std                     0.394031      0.660456      0.756164             NaN   
min                     1.000000      1.000000      1.000000             NaN   
25%                     1.000000      3.000000      2.000000             NaN   
50%                     1.000000      3.000000      2.000000             NaN   
75%                     1.000000      3.000000      3.000000             NaN   
max                     5.000000      5.000000      5.000000             NaN   

       Body Type Position Height  Weight     LS     ST     RS     LW     LF  \
count      18159    18147  18159   18159  16122  16122  16122  16122  16122   
unique        10       27     21      57     93     93     93    105    102   
top       Normal       ST    6'0  165lbs   61+2   61+2   61+2   63+2   61+2   
freq       10595     2152   2881    1483    705    705    705    745    713   
mean         NaN      NaN    NaN     NaN    NaN    NaN    NaN    NaN    NaN   
std          NaN      NaN    NaN     NaN    NaN    NaN    NaN    NaN    NaN   
min          NaN      NaN    NaN     NaN    NaN    NaN    NaN    NaN    NaN   
25%          NaN      NaN    NaN     NaN    NaN    NaN    NaN    NaN    NaN   
50%          NaN      NaN    NaN     NaN    NaN    NaN    NaN    NaN    NaN   
75%          NaN      NaN    NaN     NaN    NaN    NaN    NaN    NaN    NaN   
max          NaN      NaN    NaN     NaN    NaN    NaN    NaN    NaN    NaN   

           CF     RF     RW    LAM    CAM    RAM     LM    LCM     CM    RCM  \
count   16122  16122  16122  16122  16122  16122  16122  16122  16122  16122   
unique    102    102    105    101    101    101    100     92     92     92   
top      61+2   61+2   63+2   61+2   61+2   61+2   61+2   58+2   58+2   58+2   
freq      713    713    745    744 

### Checking "Age"

In [8]:
data["Age"].isnull().sum()

0

### Checking Nationality

In [9]:
data["Nationality"].isnull().sum()

0

### Checking Overall

In [10]:
data["Overall"].isnull().sum()

0

### Checking Potential

In [11]:
data["Potential"].isnull().sum()

0

### Checking Value

In [12]:
data["Value"].isnull().sum()

0

### Checking Wage

In [13]:
data["Wage"].isnull().sum()

0

### Checking Special

In [14]:
data["Special"].isnull().sum()

0

### Checking Preferred Foot

In [15]:
data["Preferred Foot"].isnull().sum()

48

Only 48 of such rows have a NaN value for "Preferred Foot", will drop these rows as they are insignificant

#### Creating Index to drop for Preferred Foot variable

In [16]:
index_to_drop_pf=[]

for i in range(len(data["Preferred Foot"])):
    if type(data["Preferred Foot"][i]) != str:
        index_to_drop_pf.append(i)
        
index_to_drop_pf

[245,
 337,
 580,
 1247,
 1688,
 2916,
 3925,
 3973,
 4706,
 5087,
 6439,
 6804,
 6932,
 7272,
 7521,
 7945,
 8610,
 9013,
 9014,
 9739,
 10136,
 10217,
 10502,
 11742,
 11885,
 12206,
 12755,
 13066,
 13946,
 14594,
 14796,
 15121,
 15229,
 15902,
 16035,
 16278,
 16823,
 16827,
 16830,
 17160,
 17163,
 17164,
 17165,
 17167,
 17788,
 18013,
 18018,
 18138]

### Checking International Reputation

In [17]:
data["International Reputation"].isnull().sum()

48

#### Creating Index to drop for International Reputation variable

In [18]:
type(data["International Reputation"][3])

numpy.float64

In [19]:
index_to_drop_ir_tuple=np.where(data["International Reputation"].isnull())

type(index_to_drop_ir_tuple)

tuple

In [20]:
index_to_drop_ir = []

for i in index_to_drop_ir_tuple:
    index_to_drop_ir.append(i)
    
index_to_drop_ir = np.asarray(index_to_drop_ir)

index_to_drop_ir = index_to_drop_ir.flatten()

index_to_drop_ir = index_to_drop_ir.tolist()

index_to_drop_ir

[245,
 337,
 580,
 1247,
 1688,
 2916,
 3925,
 3973,
 4706,
 5087,
 6439,
 6804,
 6932,
 7272,
 7521,
 7945,
 8610,
 9013,
 9014,
 9739,
 10136,
 10217,
 10502,
 11742,
 11885,
 12206,
 12755,
 13066,
 13946,
 14594,
 14796,
 15121,
 15229,
 15902,
 16035,
 16278,
 16823,
 16827,
 16830,
 17160,
 17163,
 17164,
 17165,
 17167,
 17788,
 18013,
 18018,
 18138]

Only 48 of such rows have a NaN value for "Preferred Foot", will drop these rows as they are insignificant

### Checking Weak Foot

In [21]:
data["Weak Foot"].isnull().sum()

48

#### Creating index to drop for Weak Foot variable

In [22]:
index_to_drop_wf_tuple=np.where(data["Weak Foot"].isnull())

index_to_drop_wf = []

for i in index_to_drop_wf_tuple:
    index_to_drop_wf.append(i)
    
index_to_drop_wf = np.asarray(index_to_drop_wf)

index_to_drop_wf = index_to_drop_wf.flatten()

index_to_drop_wf = index_to_drop_wf.tolist()

### Checking Skill Moves

In [23]:
data["Skill Moves"].isnull().sum()

48

#### Creating index to drop for Skill Moves variable

In [24]:
index_to_drop_sm_tuple=np.where(data["Skill Moves"].isnull())

index_to_drop_sm = []

for i in index_to_drop_sm_tuple:
    index_to_drop_sm.append(i)
    
index_to_drop_sm = np.asarray(index_to_drop_sm)

index_to_drop_sm = index_to_drop_sm.flatten()

index_to_drop_sm = index_to_drop_sm.tolist()

### Checking Work Rate

In [25]:
data["Work Rate"].isnull().sum()

48

In [26]:
index_to_drop_wr_tuple=np.where(data["Work Rate"].isnull())

index_to_drop_wr = []

for i in index_to_drop_wr_tuple:
    index_to_drop_wr.append(i)
    
index_to_drop_wr = np.asarray(index_to_drop_wr)

index_to_drop_wr = index_to_drop_wr.flatten()

index_to_drop_wr = index_to_drop_wr.tolist()

### Checking Body Type

In [27]:
data["Body Type"].isnull().sum()

48

In [28]:
index_to_drop_bt_tuple=np.where(data["Body Type"].isnull())

index_to_drop_bt = []

for i in index_to_drop_bt_tuple:
    index_to_drop_bt.append(i)
    
index_to_drop_bt = np.asarray(index_to_drop_bt)

index_to_drop_bt = index_to_drop_bt.flatten()

index_to_drop_bt = index_to_drop_bt.tolist()

Something seems wrong with Body Type, there are supposed to be 3 unique values from what I understand, but there seems to be more. Shall check on this

In [29]:
data["Body Type"].unique()

array(['Lean', 'Normal', 'Stocky', 'C. Ronaldo', nan, 'Akinfenwa',
       'Messi', 'Neymar', 'Courtois', 'Shaqiri', 'PLAYER_BODY_TYPE_25'],
      dtype=object)

Seems that "Akinfenwa", "C. Ronaldo", "Courtois", "Messi", "Neymar", "PLAYER_BODY_TYPE_25" & "Shaqiri" are unique values that are for specific rows only. The 3 main type of body is "Lean", "Normal" & "Stocky".

As such we will only adopt and encode the 3 main types and drop the rest.

In [30]:
akinfenwa = np.where(data["Body Type"]=="Akinfenwa")
ronaldo = np.where(data["Body Type"]=="C. Ronaldo")
messi = np.where(data["Body Type"]=="Messi")
neymar = np.where(data["Body Type"]=="Neymar")
courtois = np.where(data["Body Type"]=="Courtois")
shaqiri = np.where(data["Body Type"]=="Shaqiri")
p25 = np.where(data["Body Type"]=="PLAYER_BODY_TYPE_25")

In [31]:
index_to_drop_bt_weird = np.concatenate((akinfenwa, ronaldo, messi, neymar, courtois, shaqiri, p25))

index_to_drop_bt_weird

    
index_to_drop_bt_weird = np.asarray(index_to_drop_bt_weird)

index_to_drop_bt_weird = index_to_drop_bt_weird.flatten()

index_to_drop_bt_weird = index_to_drop_bt_weird.tolist()

index_to_drop_bt_weird

[637, 45, 699, 3255, 3412, 3668, 6411]

In [32]:
index_to_drop_bt_overall = list(set().union(index_to_drop_bt, index_to_drop_bt_weird))

index_to_drop_bt_overall.sort()

print(len(index_to_drop_bt_overall))

print(index_to_drop_bt_overall)

55
[45, 245, 337, 580, 637, 699, 1247, 1688, 2916, 3255, 3412, 3668, 3925, 3973, 4706, 5087, 6411, 6439, 6804, 6932, 7272, 7521, 7945, 8610, 9013, 9014, 9739, 10136, 10217, 10502, 11742, 11885, 12206, 12755, 13066, 13946, 14594, 14796, 15121, 15229, 15902, 16035, 16278, 16823, 16827, 16830, 17160, 17163, 17164, 17165, 17167, 17788, 18013, 18018, 18138]


### Checking Position

In [33]:
data["Position"].isnull().sum()

60

In [34]:
data["Position"].unique()

array(['RCM', 'LF', 'CM', 'LCM', 'CDM', 'LCB', 'GK', 'ST', 'LM', 'CB',
       'CAM', 'RM', 'RDM', 'RS', 'LB', 'LS', 'RCB', 'RW', 'LW', 'RB', nan,
       'RWB', 'LWB', 'LDM', 'CF', 'RF', 'LAM', 'RAM'], dtype=object)

In [35]:
index_to_drop_p_tuple=np.where(data["Position"].isnull())

index_to_drop_p = []

for i in index_to_drop_p_tuple:
    index_to_drop_p.append(i)
    
index_to_drop_p = np.asarray(index_to_drop_p)

index_to_drop_p = index_to_drop_p.flatten()

index_to_drop_p = index_to_drop_p.tolist()

### Checking Height

In [36]:
data["Height"].isnull().sum()

48

In [37]:
index_to_drop_h_tuple=np.where(data["Height"].isnull())

index_to_drop_h = []

for i in index_to_drop_h_tuple:
    index_to_drop_h.append(i)
    
index_to_drop_h = np.asarray(index_to_drop_h)

index_to_drop_h = index_to_drop_h.flatten()

index_to_drop_h = index_to_drop_h.tolist()

### Checking Weight

In [38]:
data["Weight"].isnull().sum()

48

In [39]:
index_to_drop_w_tuple=np.where(data["Weight"].isnull())

index_to_drop_w = []

for i in index_to_drop_w_tuple:
    index_to_drop_w.append(i)
    
index_to_drop_w = np.asarray(index_to_drop_w)

index_to_drop_w = index_to_drop_w.flatten()

index_to_drop_w = index_to_drop_w.tolist()

### Checking LS

In [40]:
data["LS"].isnull().sum()

2085

In [41]:
GK=data["Position"]=="GK"
GK.sum()

2025

In [42]:
2085-2025

60

##### _Seems like for the individual positional ratings like "LS", "ST", all the way to "RB", they will all have "GK" as missing values (all of them have the same number of values, 16122) and also (assumption) the 60 values which is missing in the data["Position"] variable. As such we will not be continuing the check on these positional rating variables. We will later on map all the outfield players to have 0 for "GK" positional rating and the opposite for all GK players._

### Checking Crossing

In [43]:
data["Crossing"].isnull().sum()

48

In [44]:
index_to_drop_cross_tuple=np.where(data["Crossing"].isnull())

index_to_drop_cross = []

for i in index_to_drop_cross_tuple:
    index_to_drop_cross.append(i)
    
index_to_drop_cross = np.asarray(index_to_drop_cross)

index_to_drop_cross = index_to_drop_cross.flatten()

index_to_drop_cross = index_to_drop_cross.tolist()

### Checking Finishing

In [45]:
data["Finishing"].isnull().sum()

48

In [46]:
index_to_drop_finish_tuple=np.where(data["Finishing"].isnull())

index_to_drop_finish = []

for i in index_to_drop_finish_tuple:
    index_to_drop_finish.append(i)
    
index_to_drop_finish = np.asarray(index_to_drop_finish)

index_to_drop_finish = index_to_drop_finish.flatten()

index_to_drop_finish = index_to_drop_finish.tolist()

### Checking HeadingAccuracy

In [47]:
data["HeadingAccuracy"].isnull().sum()

48

In [48]:
index_to_drop_head_tuple=np.where(data["HeadingAccuracy"].isnull())

index_to_drop_head = []

for i in index_to_drop_head_tuple:
    index_to_drop_head.append(i)
    
index_to_drop_head = np.asarray(index_to_drop_head)

index_to_drop_head = index_to_drop_head.flatten()

index_to_drop_head = index_to_drop_head.tolist()

### Checking ShortPassing

In [49]:
data["ShortPassing"].isnull().sum()

48

In [50]:
index_to_drop_shortp_tuple=np.where(data["ShortPassing"].isnull())

index_to_drop_shortp = []

for i in index_to_drop_shortp_tuple:
    index_to_drop_shortp.append(i)
    
index_to_drop_shortp = np.asarray(index_to_drop_shortp)

index_to_drop_shortp = index_to_drop_shortp.flatten()

index_to_drop_shortp = index_to_drop_shortp.tolist()

### Checking Volleys

In [51]:
data["Volleys"].isnull().sum()

48

In [52]:
index_to_drop_volley_tuple=np.where(data["Volleys"].isnull())

index_to_drop_volley = []

for i in index_to_drop_volley_tuple:
    index_to_drop_volley.append(i)
    
index_to_drop_volley = np.asarray(index_to_drop_volley)

index_to_drop_volley = index_to_drop_volley.flatten()

index_to_drop_volley = index_to_drop_volley.tolist()

### Checking Dribbling

In [53]:
data["Dribbling"].isnull().sum()

48

In [54]:
index_to_drop_drib_tuple=np.where(data["Dribbling"].isnull())

index_to_drop_drib = []

for i in index_to_drop_drib_tuple:
    index_to_drop_drib.append(i)
    
index_to_drop_drib = np.asarray(index_to_drop_drib)

index_to_drop_drib = index_to_drop_drib.flatten()

index_to_drop_drib = index_to_drop_drib.tolist()

### Checking Curve

In [55]:
data["Curve"].isnull().sum()

48

### Checking FKAccuracy

In [56]:
data["FKAccuracy"].isnull().sum()

48

In [57]:
index_to_drop_FK_tuple=np.where(data["FKAccuracy"].isnull())

index_to_drop_FK = []

for i in index_to_drop_FK_tuple:
    index_to_drop_FK.append(i)
    
index_to_drop_FK = np.asarray(index_to_drop_FK)

index_to_drop_FK = index_to_drop_FK.flatten()

index_to_drop_FK = index_to_drop_FK.tolist()

### Checking LongPassing

In [58]:
data["LongPassing"].isnull().sum()

48

In [59]:
index_to_drop_LP_tuple=np.where(data["LongPassing"].isnull())

index_to_drop_LP = []

for i in index_to_drop_LP_tuple:
    index_to_drop_LP.append(i)
    
index_to_drop_LP = np.asarray(index_to_drop_LP)

index_to_drop_LP = index_to_drop_LP.flatten()

index_to_drop_LP = index_to_drop_LP.tolist()

### Checking BallControl

In [60]:
data["BallControl"].isnull().sum()

48

In [61]:
index_to_drop_BC_tuple=np.where(data["BallControl"].isnull())

index_to_drop_BC = []

for i in index_to_drop_BC_tuple:
    index_to_drop_BC.append(i)
    
index_to_drop_BC = np.asarray(index_to_drop_BC)

index_to_drop_BC = index_to_drop_BC.flatten()

index_to_drop_BC = index_to_drop_BC.tolist()

### Checking Acceleration

In [62]:
data["Acceleration"].isnull().sum()

48

In [63]:
index_to_drop_accel_tuple=np.where(data["Acceleration"].isnull())

index_to_drop_accel = []

for i in index_to_drop_accel_tuple:
    index_to_drop_accel.append(i)
    
index_to_drop_accel = np.asarray(index_to_drop_accel)

index_to_drop_accel = index_to_drop_accel.flatten()

index_to_drop_accel = index_to_drop_accel.tolist()

### Checking SprintSpeed

In [64]:
data["SprintSpeed"].isnull().sum()

48

In [65]:
index_to_drop_SS_tuple=np.where(data["SprintSpeed"].isnull())

index_to_drop_SS = []

for i in index_to_drop_SS_tuple:
    index_to_drop_SS.append(i)
    
index_to_drop_SS = np.asarray(index_to_drop_SS)

index_to_drop_SS = index_to_drop_SS.flatten()

index_to_drop_SS = index_to_drop_SS.tolist()

### Checking Agility

In [66]:
data["Agility"].isnull().sum()

48

In [67]:
index_to_drop_agi_tuple=np.where(data["Agility"].isnull())

index_to_drop_agi = []

for i in index_to_drop_agi_tuple:
    index_to_drop_agi.append(i)
    
index_to_drop_agi = np.asarray(index_to_drop_agi)

index_to_drop_agi = index_to_drop_agi.flatten()

index_to_drop_agi = index_to_drop_agi.tolist()

### Checking Reactions

In [68]:
data["Reactions"].isnull().sum()

48

In [69]:
index_to_drop_react_tuple=np.where(data["Reactions"].isnull())

index_to_drop_react = []

for i in index_to_drop_react_tuple:
    index_to_drop_react.append(i)
    
index_to_drop_react = np.asarray(index_to_drop_react)

index_to_drop_react = index_to_drop_react.flatten()

index_to_drop_react = index_to_drop_react.tolist()

### Checking Balance

In [70]:
data["Balance"].isnull().sum()

48

In [71]:
index_to_drop_bal_tuple=np.where(data["Balance"].isnull())

index_to_drop_bal = []

for i in index_to_drop_bal_tuple:
    index_to_drop_bal.append(i)
    
index_to_drop_bal = np.asarray(index_to_drop_bal)

index_to_drop_bal = index_to_drop_bal.flatten()

index_to_drop_bal = index_to_drop_bal.tolist()

### Checking ShotPower

In [72]:
data["ShotPower"].isnull().sum()

48

In [73]:
index_to_drop_shotp_tuple=np.where(data["ShotPower"].isnull())

index_to_drop_shotp = []

for i in index_to_drop_shotp_tuple:
    index_to_drop_shotp.append(i)
    
index_to_drop_shotp = np.asarray(index_to_drop_shotp)

index_to_drop_shotp = index_to_drop_shotp.flatten()

index_to_drop_shotp = index_to_drop_shotp.tolist()

### Checking Jumping

In [74]:
data["Jumping"].isnull().sum()

48

In [75]:
index_to_drop_jump_tuple=np.where(data["Jumping"].isnull())

index_to_drop_jump = []

for i in index_to_drop_jump_tuple:
    index_to_drop_jump.append(i)
    
index_to_drop_jump = np.asarray(index_to_drop_jump)

index_to_drop_jump = index_to_drop_jump.flatten()

index_to_drop_jump = index_to_drop_jump.tolist()

### Checking Stamina

In [76]:
data["Stamina"].isnull().sum()

48

In [77]:
index_to_drop_stam_tuple=np.where(data["Stamina"].isnull())

index_to_drop_stam = []

for i in index_to_drop_stam_tuple:
    index_to_drop_stam.append(i)
    
index_to_drop_stam = np.asarray(index_to_drop_stam)

index_to_drop_stam = index_to_drop_stam.flatten()

index_to_drop_stam = index_to_drop_stam.tolist()

### Checking Stength

In [78]:
data["Strength"].isnull().sum()

48

In [79]:
index_to_drop_str_tuple=np.where(data["Strength"].isnull())

index_to_drop_str = []

for i in index_to_drop_str_tuple:
    index_to_drop_str.append(i)
    
index_to_drop_str = np.asarray(index_to_drop_str)

index_to_drop_str = index_to_drop_str.flatten()

index_to_drop_str = index_to_drop_str.tolist()

### Checking LongShots

In [80]:
data["LongShots"].isnull().sum()

48

In [81]:
index_to_drop_lshot_tuple=np.where(data["LongShots"].isnull())

index_to_drop_lshot = []

for i in index_to_drop_lshot_tuple:
    index_to_drop_lshot.append(i)
    
index_to_drop_lshot = np.asarray(index_to_drop_lshot)

index_to_drop_lshot = index_to_drop_lshot.flatten()

index_to_drop_lshot = index_to_drop_lshot.tolist()

### Checking Aggression

In [82]:
data["Aggression"].isnull().sum()

48

In [83]:
index_to_drop_agg_tuple=np.where(data["Aggression"].isnull())

index_to_drop_agg = []

for i in index_to_drop_agg_tuple:
    index_to_drop_agg.append(i)
    
index_to_drop_agg = np.asarray(index_to_drop_agg)

index_to_drop_agg = index_to_drop_agg.flatten()

index_to_drop_agg = index_to_drop_agg.tolist()

### Checking Interceptions

In [84]:
data["Interceptions"].isnull().sum()

48

In [85]:
index_to_drop_inter_tuple=np.where(data["Interceptions"].isnull())

index_to_drop_inter = []

for i in index_to_drop_inter_tuple:
    index_to_drop_inter.append(i)
    
index_to_drop_inter = np.asarray(index_to_drop_inter)

index_to_drop_inter = index_to_drop_inter.flatten()

index_to_drop_inter = index_to_drop_inter.tolist()

### Checking Positioning

In [86]:
data["Positioning"].isnull().sum()

48

In [87]:
index_to_drop_pos_tuple=np.where(data["Positioning"].isnull())

index_to_drop_pos = []

for i in index_to_drop_pos_tuple:
    index_to_drop_pos.append(i)
    
index_to_drop_pos = np.asarray(index_to_drop_pos)

index_to_drop_pos = index_to_drop_pos.flatten()

index_to_drop_pos = index_to_drop_pos.tolist()

### Checking Vision

In [88]:
data["Vision"].isnull().sum()

48

In [89]:
index_to_drop_vis_tuple=np.where(data["Vision"].isnull())

index_to_drop_vis = []

for i in index_to_drop_vis_tuple:
    index_to_drop_vis.append(i)
    
index_to_drop_vis = np.asarray(index_to_drop_vis)

index_to_drop_vis = index_to_drop_vis.flatten()

index_to_drop_vis = index_to_drop_vis.tolist()

### Checking Penalties

In [90]:
data["Penalties"].isnull().sum()

48

In [91]:
index_to_drop_pen_tuple=np.where(data["Penalties"].isnull())

index_to_drop_pen = []

for i in index_to_drop_pen_tuple:
    index_to_drop_pen.append(i)
    
index_to_drop_pen = np.asarray(index_to_drop_pen)

index_to_drop_pen = index_to_drop_pen.flatten()

index_to_drop_pen = index_to_drop_pen.tolist()

### Checking Composure

In [92]:
data["Composure"].isnull().sum()

48

In [93]:
index_to_drop_comp_tuple=np.where(data["Composure"].isnull())

index_to_drop_comp = []

for i in index_to_drop_comp_tuple:
    index_to_drop_comp.append(i)
    
index_to_drop_comp = np.asarray(index_to_drop_comp)

index_to_drop_comp = index_to_drop_comp.flatten()

index_to_drop_comp = index_to_drop_comp.tolist()

### Checking Marking

In [94]:
data["Marking"].isnull().sum()

48

In [95]:
index_to_drop_mark_tuple=np.where(data["Marking"].isnull())

index_to_drop_mark = []

for i in index_to_drop_mark_tuple:
    index_to_drop_mark.append(i)
    
index_to_drop_mark = np.asarray(index_to_drop_mark)

index_to_drop_mark = index_to_drop_mark.flatten()

index_to_drop_mark = index_to_drop_mark.tolist()

### Checking StandingTackle

In [96]:
data["StandingTackle"].isnull().sum()

48

In [97]:
index_to_drop_standt_tuple=np.where(data["StandingTackle"].isnull())

index_to_drop_standt = []

for i in index_to_drop_standt_tuple:
    index_to_drop_standt.append(i)
    
index_to_drop_standt = np.asarray(index_to_drop_standt)

index_to_drop_standt = index_to_drop_standt.flatten()

index_to_drop_standt = index_to_drop_standt.tolist()

### Checking SlidingTackle

In [98]:
data["SlidingTackle"].isnull().sum()

48

In [99]:
index_to_drop_slidet_tuple=np.where(data["SlidingTackle"].isnull())

index_to_drop_slidet = []

for i in index_to_drop_slidet_tuple:
    index_to_drop_slidet.append(i)
    
index_to_drop_slidet = np.asarray(index_to_drop_slidet)

index_to_drop_slidet = index_to_drop_slidet.flatten()

index_to_drop_slidet = index_to_drop_slidet.tolist()

### Checking GKDiving

In [100]:
data["GKDiving"].isnull().sum()

48

In [101]:
index_to_drop_GKD_tuple=np.where(data["GKDiving"].isnull())

index_to_drop_GKD = []

for i in index_to_drop_GKD_tuple:
    index_to_drop_GKD.append(i)
    
index_to_drop_GKD = np.asarray(index_to_drop_GKD)

index_to_drop_GKD = index_to_drop_GKD.flatten()

index_to_drop_GKD = index_to_drop_GKD.tolist()

### Checking GKHandling

In [102]:
data["GKHandling"].isnull().sum()

48

In [103]:
index_to_drop_GKH_tuple=np.where(data["GKHandling"].isnull())

index_to_drop_GKH = []

for i in index_to_drop_GKH_tuple:
    index_to_drop_GKH.append(i)
    
index_to_drop_GKH = np.asarray(index_to_drop_GKH)

index_to_drop_GKH = index_to_drop_GKH.flatten()

index_to_drop_GKH = index_to_drop_GKH.tolist()

### Checking GKKicking

In [104]:
data["GKKicking"].isnull().sum()

48

In [105]:
index_to_drop_GKK_tuple=np.where(data["GKKicking"].isnull())

index_to_drop_GKK = []

for i in index_to_drop_GKK_tuple:
    index_to_drop_GKK.append(i)
    
index_to_drop_GKK = np.asarray(index_to_drop_GKK)

index_to_drop_GKK = index_to_drop_GKK.flatten()

index_to_drop_GKK = index_to_drop_GKK.tolist()

### Checking GKPositioning

In [106]:
data["GKPositioning"].isnull().sum()

48

In [107]:
index_to_drop_GKP_tuple=np.where(data["GKPositioning"].isnull())

index_to_drop_GKP = []

for i in index_to_drop_GKP_tuple:
    index_to_drop_GKP.append(i)
    
index_to_drop_GKP = np.asarray(index_to_drop_GKP)

index_to_drop_GKP = index_to_drop_GKP.flatten()

index_to_drop_GKP = index_to_drop_GKP.tolist()

### Checking GKReflexes

In [108]:
data["GKReflexes"].isnull().sum()

48

In [109]:
index_to_drop_GKR_tuple=np.where(data["GKReflexes"].isnull())

index_to_drop_GKR = []

for i in index_to_drop_GKR_tuple:
    index_to_drop_GKR.append(i)
    
index_to_drop_GKR = np.asarray(index_to_drop_GKR)

index_to_drop_GKR = index_to_drop_GKR.flatten()

index_to_drop_GKR = index_to_drop_GKR.tolist()

### Checking Release Clause

In [110]:
data["Release Clause"].isnull().sum()

1564

In [111]:
index_to_drop_RC_tuple=np.where(data["Release Clause"].isnull())

index_to_drop_RC = []

for i in index_to_drop_RC_tuple:
    index_to_drop_RC.append(i)
    
index_to_drop_RC = np.asarray(index_to_drop_RC)

index_to_drop_RC = index_to_drop_RC.flatten()

index_to_drop_RC = index_to_drop_RC.tolist()

index_to_drop_RC

[12,
 56,
 85,
 106,
 129,
 145,
 245,
 312,
 337,
 365,
 427,
 471,
 482,
 580,
 581,
 585,
 624,
 630,
 642,
 648,
 660,
 719,
 778,
 785,
 857,
 902,
 907,
 912,
 924,
 972,
 994,
 1006,
 1023,
 1030,
 1046,
 1091,
 1111,
 1124,
 1125,
 1155,
 1227,
 1234,
 1247,
 1249,
 1270,
 1299,
 1340,
 1381,
 1403,
 1435,
 1455,
 1460,
 1467,
 1506,
 1508,
 1513,
 1514,
 1517,
 1525,
 1589,
 1598,
 1675,
 1685,
 1688,
 1698,
 1699,
 1780,
 1808,
 1821,
 1838,
 1847,
 1872,
 1908,
 1912,
 1922,
 1935,
 1958,
 1979,
 1984,
 2004,
 2025,
 2045,
 2072,
 2076,
 2082,
 2091,
 2093,
 2110,
 2114,
 2131,
 2133,
 2135,
 2140,
 2166,
 2251,
 2252,
 2260,
 2271,
 2292,
 2333,
 2342,
 2348,
 2353,
 2365,
 2370,
 2397,
 2405,
 2410,
 2415,
 2416,
 2420,
 2449,
 2453,
 2460,
 2483,
 2494,
 2501,
 2504,
 2545,
 2551,
 2553,
 2554,
 2603,
 2622,
 2632,
 2640,
 2646,
 2647,
 2650,
 2668,
 2670,
 2673,
 2701,
 2724,
 2736,
 2834,
 2840,
 2841,
 2848,
 2850,
 2863,
 2888,
 2916,
 2918,
 2920,
 2928,
 2979,
 2988

## Comments

(1) Seems that most of the data have the same 48 missing variables, as such we will opt to drop these 48

(2) "Body Type" variable has some erroneous values in the form of names as body types, we will drop these values together with the 48 in (1)

(3) For positional ratings, we will not drop the "extra" missing values. Instead we will map it later on - outfield players will have 0 for "GK" & Goalkeepers will have 0 for outfield positional ratings

(4) As for "Release Clause", some players just do not have release clauses in their contracts (eg. most BPL players I know do not have it). As such we will not drop the players without release clauses. Instead we will assign the average value of the release clauses to the players who do not have it. We will then be able to see if having a higher than average release clause will affect the player potential.

## Combining the list of index to drop